Style engine: pre-core migration documentation and house keeping

ramonjd opened the pull request:
## What?

Adding further documentation, cleaning up change log ready to populate with real changes  (TODO)  :P

Removing trailing references to `'enqueue' => true,`

Updating PHP unit tests to conform with WordPress PHP test standards: https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/


## Why?
Getting the style engine package into a reasonable state to migrate to core for 6.1.

See backport PR: https://github.com/WordPress/wordpress-develop/pull/3199

## TODO
- [x] Merge https://github.com/WordPress/gutenberg/pull/43886 after it lands

## Testing Instructions

- Proofreading 👓 
- Tests pass 🍏 
- Linting ✏️ 
- Check that `wp.styleEngine.compileCSS` and `wp.styleEngine.getCSSRules` are available in the global window scope in the editor.

ramonjd committed:
Initial commit. Adding further documentation, cleaning up changelog ready to populate with real changes :P

ramonjd committed:
Typos, formatting

ramonjd committed:
Typos, formatting

ramonjd committed:
Sorting out change log and formatting

ramonjd commented on tools/webpack/packages.js:
```diff
@@ -23,11 +23,7 @@ const WORDPRESS_NAMESPACE = '@wordpress/';
-	'@wordpress/interface',
-	'@wordpress/style-engine',
-];
+const BUNDLED_PACKAGES = [ '@wordpress/icons', '@wordpress/interface' ];
```
@youknowriad Do you remember? Is this the only change required to ensure that the JS methods appear in the global window scope? E.g., `wp.styleEngine`

I couldn't seem to get it working.

gziolo replied:
It needs to get removed in all places where `@wordpress/style-engine` is explicitly marked as an exception like here:
 
https://github.com/WordPress/gutenberg/blob/ce4feda1038708b563b7f98ae6b886671dc56945/packages/dependency-extraction-webpack-plugin/lib/util.js#L2-L6

Well, it's probably the only place left 😄

aristath commented on packages/style-engine/docs/using-the-style-engine-with-block-supports.md:
```diff
@@ -0,0 +1,156 @@
+
+
+
+
```
No need for multiple empty lines 👍 

```suggestion

```

ramonjd committed:
Update packages/style-engine/docs/using-the-style-engine-with-block-supports.md

No empty lines! :D

Co-authored-by: Ari Stathopoulos <aristath@gmail.com>

ramonjd committed:
Update README.md

typo

ramonjd committed:
Remove enqueue option. It does nothing.

andrewserong commented on packages/style-engine/docs/using-the-style-engine-with-block-supports.md:
```diff
@@ -0,0 +1,151 @@
+# Using the Style Engine to generate block supports styles
```
Is the intention that this document is to be available in the developer handbook? If so, I _think_ we might need to add an entry in the `docs/manifest.json` file for it. Here's the line for the parent style engine readme: https://github.com/WordPress/gutenberg/blob/trunk/docs/manifest.json#L1815

From the looks of it, we can probably create a child page for this document to live in by copying one of the entries from the components package, e.g. [here's the entry for Notice](https://github.com/WordPress/gutenberg/blob/3b4526a7a91b6ec38d6bb067db650679ec9e3573/docs/manifest.json#L1053)

I imagine it's a matter of setting a title and a slug, the correct `markdown_source` and then setting `parent` to `"packages-style-engine"`?

ramonjd replied:
Thanks @andrewserong! Yes, I think it should be part of the docs. On it...

andrewserong commented on packages/style-engine/README.md:
```diff
@@ -30,8 +26,9 @@ See [Tracking: Add a Style Engine to manage rendering block styles #38167](https
-.
+the [top level styles in theme.json](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/).
+
+See also [Using the Style Engine to generate block supports styles.](docs/using-the-style-engine-with-block-supports.md)
```
See the other comment about where this file ends up being located in the developer handbook. We can probably figure out the final URL based on slugs. I _think_ it might wind up being something like: https://developer.wordpress.org/block-editor/reference-guides/packages/packages-style-engine/using-the-style-engine-with-block-supports ?

andrewserong commented:
This is looking good to me @ramonjd, thanks for all the house keeping here! Just left a couple of comments about how we can expose the added document. The only issue I ran into is that I couldn't access the `wp.styleEngine` object globally, but I'm not too sure how that's meant to hook together 🤔

ramonjd committed:
Updating manifest.json to include new style engine docs
Splitting the global functions out of the class files
Checking for the directory when loading rather than each individual file

ramonjd commented:
> The only issue I ran into is that I couldn't access the wp.styleEngine object globally, but I'm not too sure how that's meant to hook together

I'm not sure either 🤔 I tried a bunch of stuff, and compared the package with other packages but can't see anything that stands out. I must be missing something.

Maybe @gziolo has a few tips for us 😄

ramonjd committed:
Adding type annotation

ramonjd committed:
Updating CHANGELOG.md

gziolo commented on packages/style-engine/package.json:
```diff
@@ -1,13 +1,14 @@
 	"name": "@wordpress/style-engine",
-	"version": "0.15.0",
-	"description": "WordPress Style engine.",
+	"version": "1.0.0",
```
```suggestion
	"version": "1.0.0-prerelease",
```

Otherwise it would get bumped to `1.1.0` during npm publishing 😄

ramonjd replied:
> Otherwise it would get bumped to 1.1.0 during npm publishing

Ha, good call! Thanks for the tip 😉

gziolo commented on packages/style-engine/src/index.ts:
```diff
@@ -17,10 +17,10 @@ import { styleDefinitions } from './styles';
  *
- * @param  style   Style object, for example, the value of a block's attributes.style object or the top level styles in theme.json
- * @param  options Options object with settings to adjust how the styles are generated.
+ * @param {Style}        style   Style object, for example, the value of a block's attributes.style object or the top level styles in theme.json
```
Out of curiosity, is the type for `@param` enforced in TypeScript files, too? I had an impression this was lifted as we can infer the type from TS syntax.

ramonjd replied:
You're right: there's a bit of tautology here. I was think about folks who were reading the code.

Actually I see the docs are doubling up, so it's probably safe to revert these. Thanks!

ramonjd committed:
Updated version to prerelease
Reverted JS comments for TS types and rebuilt docs
Removed style engine from bundle exception array

ramonjd committed:
Rebase after #43886

ramonjd commented on phpunit/script-loader.php:
```diff
@@ -39,7 +39,6 @@ public function test_enqueue_stored_styles() {
 			$core_styles_to_enqueue,
 			array(
 				'context' => 'block-supports',
-				'enqueue' => true,
```
The `enqueue` option doesn't do anything and is deprecated.

gziolo commented on packages/dependency-extraction-webpack-plugin/lib/util.js:
```diff
@@ -1,9 +1,5 @@
-	'@wordpress/interface',
-	'@wordpress/style-engine',
-];
+const BUNDLED_PACKAGES = [ '@wordpress/icons', '@wordpress/interface' ];
```
I have just realized that is probably a breaking change from the perspective of the webpack plugin when developing for WordPress that still doesn't have `wp.styleEngine` exposed. I didn't think about it before. It will work correctly only when the Gutenberg plugin is enabled until WordPress 6.1 is out.

ramonjd replied:
Oh, good call. Thanks for spotting. 

Should I revert it for now?

gziolo replied:
Actually, it might not be an issue at all if people don't reference the `@wordpress/style-engine` in their code. How likely it is, that plugin would use this new package? Maybe, it's enough to leave a note in both the package and the webpack plugin that it will work with WordPress 6.1 and above.

ramonjd replied:
> How likely it is, that plugin would use this new package? 

It's pretty limited and closely-coupled with the block editor, so pretty low, but you never know with WordPress 😄 

> Maybe, it's enough to leave a note in both the package and the webpack plugin that it will work with WordPress 6.1 and above.

I'll add this to the style engine package's README and function code comments, thanks! 🙇 

By "webpack plugin" do you mean the `DependencyExtractionWebpackPlugin`? 

Where and how would be appropriate?

gziolo replied:
`DependencyExtractionWebpackPlugin`, yes this package. It's also the context where I left the message so I assumed we talk about it :)

Maybe, we can simply add a note in the CHANGELOG file:

```
### New Features

- Include `@wordpress/style-engine` on the list of external dependencies to allow using `wp.styleEngine` global with WordPress 6.1 and beyond ([#43840](https://github.com/WordPress/gutenberg/pull/43840)).
```

ramonjd replied:
That's awesome, thank you! 🙇 

>  It's also the context where I left the message so I assumed we talk about it :)

Ha, no worries. I thought so, but I like to double check because I'm likely to add a comment in the `.nvmrc` file or somewhere else silly if I don't make sure 🙃

ramonjd committed:
Updating change logs and emphasizing that new global JS methods are available since 6.1

ramonjd committed:
Checking for functions and classes in tests so that Gutenberg tests don't fail during the backport to Core.
This is valid for 6.1. After that, developing the packages might require another solution, such as testing against `gutenberg_` prefixed functions.

ramonjd committed:
Formatting PHP doc comments and params that describe associative arrays

Updating test function names and annotations.

ramonjd committed:
Formatting

ramonjd committed:
Combining tests for get_declarations_string
Updating php doc params

ramonjd committed:
Adding message parameter to assertions where more than one assertion is used in a test method.

ramonjd committed:
Updating PHP docs
Testing for WP_Style_Engine_CSS_Rules_Store in WP_Style_Engine_Processor->add_store

ramonjd committed:
Text domain added to __()

ramonjd committed:
Refactoring sanitization tests for CSS declarations
Counting calls to mock filters executed in safecss_filter_attr() (kses.php)

ramonjd committed:
Formatting styleEngine tests
Removing duplicated test (we were already testing context below)

ramonjd committed:
My blod is typo

ramonjd committed:
Active voice rather than imperative

ramonjd committed:
Remove Gutenberg suffix

ramonjd committed:
Adding message to script loader block support enqueuing expectations

ramonjd committed:
Grammar, formatting, removing class_exists checks

ramonjd committed:
Ignore unused variable. This is a common test pattern in Core.

ramonjd committed:
Remove unusued var in tests

ockham commented:
Looks like this is pretty much ready. Would someone mind rebasing and giving their final approval so we can get land this for WP 6.1? cc/ @andrewserong @aristath 😊

ramonjd commented:
> Would someone mind rebasing

Rebased 🍺

andrewserong reviewed (approved):
Thanks so much for rebasing @ramonjd, I was just about to do that! 😄

This is looking good to me, just smoke tested that it's still all working correctly in the post and site editors, and in server rendering block supports on the site front end 👍

The style engine is also available on the global `wp` object, too:

<img width="460" alt="image" src="https://user-images.githubusercontent.com/14988353/190028799-1ca1d801-0c7b-47f4-b64c-c96eb7126211.png">

LGTM! 🎉

ockham commented:
@c4rl0sbr4v0 Let's include this on in GB 14.1 stable 🙂

cbravobernal commented:
[Cherry-picked for 14.1](https://github.com/WordPress/gutenberg/commit/9b084273203c9cec84fb67f312f8f5ef22d2982c) 🚀

bph commented:
- [Type] User Documentation is meant for End users - = UI changes that result in update on screenshots and features for content creators. Also, no changelog handling. I removed the lable. 

- [Type] Developer Documentation will go into the 'Documentation' section of a plugin release, also is meant for actual updates on the Developer documentation :-) 
- `needs dev note` is for WordPress releases and Dev notes handling/tracking. Add to the Dev Notes tracking issue for WP 6.1

ockham commented:
I just cherry-picked this PR to the wp/6.1 branch to get it included in the next release: 0fc7a572e5
